File: Course5_BackEnd(kotlin_spring)/ch02_springNkotlin/java-kotlin/src/main/java/org/example/ex05/Exam05.java
package org.example.ex05;

import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class Exam05 {
    // @FunctionalInterface
    private Predicate<String> stringPredicate = new Predicate<String>() {
        @Override
        public boolean test(String s) {
            return s.equals("?");
        }
    };

    public Exam05() {
        // 고차함수
        var strList = List.of(
                "1",
                "2",
                "홍길동",
                "함수",
                "메소드"
        );

        var result = strList.stream()
                .filter(it -> it.equals("?")) // 람다식
//                .filter(stringPredicate)
                /*.filter(new Predicate<String>() {
                    @Override
                    public boolean test(String s) {
                        return s.equals("?");
                    }
                })*/
                .collect(Collectors.toList());
    }

    public static void main(String[] args) {
        new Exam05();
    }
}

File: Course5_BackEnd(kotlin,spring)/ch01_kotilin/kotlin-example/src/main/kotlin/ex05/Exam05.kt
package ex05

import java.util.function.Predicate

fun main() {
    val numberList = listOf<Int>(1, 2, 3, 4, 5)
    val pred = Predicate<Int> { it % 2 == 0 }
/*
    val pred = object : Predicate<Int> {
        override fun test(t: Int): Boolean {
            return t % 2 == 0
        }
    }
*/

//    numberList.filter { it -> it % 2 == 0 }
    numberList.filter { it % 2 == 0 }
    numberList.stream().filter(pred)

    val add = { x:Int, y:Int -> x + y}
    val _add = fun (x: Int, y: Int): Int{
        return x + y
    }
//    println(add.invoke(2, 3))
    println(add(2, 3))
    println(_add(3, 4))

    lambda(4, 5, _add)
}

fun lambda(x: Int, y: Int, method: (Int, Int) -> Int) {
    println(method(x, y))
}
